### PR TITLE
Api to check elastic profile usage.

### PR DESCRIPTION
**Pull request by bdpiparva**
> Request

```bash
curl 'http://localhost:8153/go/api/elastic/profiles/262db87e-e34e-47cd-b509-7267d105bcf5/usages'
      -u 'admin:badger' \
      -H 'Accept: application/vnd.go.cd.v1+json' -v | jq .
```

> Response

```json 
[
  {
    "pipeline_name": "up4253fa0009-a61e-4e25-92ee-8445f038bc27",
    "stage_name": "Stage_2",
    "job_name": "Job3"
  },
  {
    "pipeline_name": "up42d44b1ffe-1d7d-437f-90f0-6c2abe4efd42",
    "stage_name": "Stage_2",
    "job_name": "Job3"
  },
  {
    "pipeline_name": "up425439409b-df8f-4192-b190-5f8095f2561f",
    "stage_name": "Stage_2",
    "job_name": "Job3"
  },
  {
    "pipeline_name": "up42aec00ae7-f971-4bb7-b374-94243e806841",
    "stage_name": "Stage_1",
    "job_name": "Job1"
  },
  {
    "pipeline_name": "up424892b750-b7f0-4ed4-a543-9938ce4797fb",
    "stage_name": "Stage_1",
    "job_name": "Job1"
  }
]
```

### CODE REVIEW

File: api/api-elastic-profile-v1/src/main/java/com/thoughtworks/go/apiv1/elasticprofile/ElasticProfileControllerV1.java [hunk 1]
```diff
@@ -75,6 +78,7 @@ public void setupRoutes() {
 
             get("", this::index);
             get(Routes.ElasticProfileAPI.ID, this::show);
+            get(Routes.ElasticProfileAPI.ID + "/usages", this::usages);
```
Comment (GaneshSPatil) [inline api/api-elastic-profile-v1/src/main/java/com/thoughtworks/go/apiv1/elasticprofile/ElasticProfileControllerV1.java]: can we add `mimeType`? This won't honor `Accept` header.

Comment (bdpiparva) [reply api/api-elastic-profile-v1/src/main/java/com/thoughtworks/go/apiv1/elasticprofile/ElasticProfileControllerV1.java]: Sure

**Comment by GaneshSPatil** [comment]
@bdpiparva -- are we waiting for something on this PR?

File: api/api-elastic-profile-v1/src/main/java/com/thoughtworks/go/apiv1/elasticprofile/ElasticProfileControllerV1.java [hunk 2]
```diff
@@ -73,11 +76,12 @@ public void setupRoutes() {
...
             before("/*", mimeType, apiAuthenticationHelper::checkAdminUserOrGroupAdminUserAnd403);
 
-            get("", this::index);
-            get(Routes.ElasticProfileAPI.ID, this::show);
-            post("", this::create);
-            put(Routes.ElasticProfileAPI.ID, this::update);
-            delete(Routes.ElasticProfileAPI.ID, this::destroy);
+            get("", mimeType, this::index);
+            get(Routes.ElasticProfileAPI.ID, mimeType, this::show);
+            get(Routes.ElasticProfileAPI.ID + "/usages", mimeType, this::usages);
```
Comment (GaneshSPatil) [inline api/api-elastic-profile-v1/src/main/java/com/thoughtworks/go/apiv1/elasticprofile/ElasticProfileControllerV1.java]: can we `/usages` under `Routes.ElasticProfileAPI`?

File: server/src/main/java/com/thoughtworks/go/server/service/ElasticProfileService.java [hunk 3]
```diff
@@ -59,4 +67,29 @@ public void create(Username currentUser, ElasticProfile elasticProfile, Localize
...
+
+    public Collection<JobConfigIdentifier> getJobsUsingElasticProfile(String profileId) {
+        if (findProfile(profileId) == null) {
+            throw new RecordNotFoundException(String.format("Elastic profile with id '%s' does not exist.", profileId));
+        }
+
+        final List<PipelineConfig> allPipelineConfigs = goConfigService.getAllPipelineConfigs();
+        final Set<JobConfigIdentifier> jobsUsingElasticProfile = new HashSet<>();
+
+        for (PipelineConfig pipelineConfig : allPipelineConfigs) {
```
Comment (GaneshSPatil) [inline server/src/main/java/com/thoughtworks/go/server/service/ElasticProfileService.java]: @bdpiparva @vishaldevgire - wouldn't this be a very slow operation? we are iterating over all-pipelines/stages/jobs and checking if a job required an elastic profile.

Can we cache this(a map of profile-to-jobs) under `GoCache` and rely on `PipelineConfigChangeListener` and `ElasticProfileChangeListener`(Need to create this) to modify cached information.

Comment (bdpiparva) [reply server/src/main/java/com/thoughtworks/go/server/service/ElasticProfileService.java]: @GaneshSPatil - Config is already cached. We tried with 4k pipelines and 200 profiles. We did not notice any lag in response.

**Pull request merged**